src/ico/decoder.rs

```diff
@@ -0,0 +1,215 @@
+use byteorder::{ReadBytesExt, LittleEndian};
+use std::io::{Read, Seek, SeekFrom};
+
+use color::ColorType;
+use image::{DecodingResult, ImageResult, ImageDecoder, ImageError};
+
+use bmp::BMPDecoder;
+use png::PNGDecoder;
+use self::InnerDecoder::*;
+
+// http://www.w3.org/TR/PNG-Structure.html
+// The first eight bytes of a PNG file always contain the following (decimal) values:
+const PNG_SIGNATURE: [u8; 8] = [137, 80, 78, 71, 13, 10, 26, 10];
+
+/// An ico decoder
+pub struct ICODecoder<R: Read> {
+    selected_entry: DirEntry,
+    inner_decoder: InnerDecoder<R>,
+}
+
+enum InnerDecoder<R: Read> {
+    BMP(BMPDecoder<R>),
+    PNG(PNGDecoder<R>)
+}
+
+#[derive(Clone, Copy, Default)]
+struct DirEntry {
+    width: u8,
+    height: u8,
+    color_count: u8,
+    reserved: u8,
+
+    num_color_planes: u16,
+    bits_per_pixel: u16,
+
+    image_length: u32,
+    image_offset: u32,
+}
+
+impl<R: Read + Seek> ICODecoder<R> {
+    /// Create a new decoder that decodes from the stream ```r```
+    pub fn new(mut r: R) -> ImageResult<ICODecoder<R>> {
+        let entries = try!(read_entries(&mut r));
+        let entry = try!(best_entry(entries));
+        let decoder = try!(entry.decoder(r));
+
+        Ok(ICODecoder {
+            selected_entry: entry,
+            inner_decoder: decoder,
+        })
+    }
+}
+
+fn read_entries<R: Read>(r: &mut R) -> ImageResult<Vec<DirEntry>> {
+    let _reserved = try!(r.read_u16::<LittleEndian>());
+    let _type = try!(r.read_u16::<LittleEndian>());
+    let count = try!(r.read_u16::<LittleEndian>());
+    (0..count).map(|_| read_entry(r)).collect()
+}
+
+fn read_entry<R: Read>(r: &mut R) -> ImageResult<DirEntry> {
+    let mut entry = DirEntry::default();
+
+    entry.width = try!(r.read_u8());
+    entry.height = try!(r.read_u8());
+    entry.color_count = try!(r.read_u8());
+    entry.reserved = try!(r.read_u8());
+
+    entry.num_color_planes = try!(r.read_u16::<LittleEndian>());
+    entry.bits_per_pixel = try!(r.read_u16::<LittleEndian>());
+
+    entry.image_length = try!(r.read_u32::<LittleEndian>());
+    entry.image_offset = try!(r.read_u32::<LittleEndian>());
+
+    Ok(entry)
+}
+
+/// Find the entry with the highest (color depth, size).
+fn best_entry(mut entries: Vec<DirEntry>) -> ImageResult<DirEntry> {
+    let mut best = try!(entries.pop().ok_or(ImageError::ImageEnd));
+    let mut best_score = (best.bits_per_pixel, best.real_width() + best.real_height());
+
+    for entry in entries.into_iter() {
+        let score = (entry.bits_per_pixel, entry.real_width() + entry.real_height());
```

Comment:
I vaguely expected multiplication here, not that it should make a big difference in practice.